File: python/rendering.py
import unicodedata

from blessings import Terminal


def _get_character_width(char: str) -> int:
    """
    As emojis are unicode characters and thus have a len(emoji) == 1, we need to check
    the unicode.east_asian_width(emoji) == "W" to consider the character as having a
    render width of 2 slots.
    """
    return 2 if unicodedata.east_asian_width(char[0]) in "W" else len(char)


def _get_rendered_character(char: str) -> str:
    """
    All characters need to be in total 3 slots wide, with whitespace added as padding.
    As emojis occupy 2 slots, we need to add just one whitespace after them.
    """
    ONE_SPACE = " "
    TWO_SPACES = "  "
    return char + TWO_SPACES if _get_character_width(char) == 1 else char + ONE_SPACE


#
# Initial rendering.
#
def parse_printable_row(
    row: list[bool], terminal: Terminal, alive_cell: str, dead_cell: str
) -> str:
    ac = _get_rendered_character(alive_cell)
    dc = _get_rendered_character(dead_cell)

    parsed_row = ""
    for val in row:
        if val:
            parsed_row += terminal.black(ac)
        else:
            parsed_row += terminal.white(dc)
    return parsed_row


def render_initial_grid(
    grid: list[list[bool]],
    terminal: Terminal,
    alive_cell: str,
    dead_cell: str,
):
    for i, row in enumerate(grid):
        with terminal.location(x=0, y=i):
            print(parse_printable_row(row, terminal, alive_cell, dead_cell))


#
# Iterative rendering, optimized to only render the diff.
#
def get_diff_for_rendering(
    old_grid: list[list[bool]], new_grid: list[list[bool]]
) -> list[tuple[int, int, bool]]:
    diff = []
    for y, row in enumerate(old_grid):
        for x, val in enumerate(row):
            if val != new_grid[y][x]:
                diff.append((x, y, new_grid[y][x]))
    return diff


def render_diff(
    diff: list[tuple[int, int, bool]],
    terminal: Terminal,
    alive_cell: str,
    dead_cell: str,
):
    ac = _get_rendered_character(alive_cell)
    dc = _get_rendered_character(dead_cell)

    for x, y, is_alive in diff:
        with terminal.location(x=x * 3, y=y):
            print(ac if is_alive else dc)

File: golang/rendering.go
package main

import (
	"fmt"
	"os"
	"os/exec"
)

// Generic utility
func ClearScreen() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
}

//
// Initial rendering
//
func GetRowForRendering(row []bool) string {
	var parsedRow string = ""
	for _, val := range row {
		if val == true {
			parsedRow += fmt.Sprint("🟪 ")
		} else {
			parsedRow += fmt.Sprint("⬜️ ")
		}
	}
	return parsedRow
}

func RenderInitialGrid(grid Grid) {
	// Terminal clear sequence from https://stackoverflow.com/a/33509850
	fmt.Printf("\033[0;0H")
	for _, row := range grid {
		fmt.Printf(GetRowForRendering(row) + "\n")
	}
}

//
// Iterative rendering, optimized to only render the diff.
//
type Diff struct {
	x, y    int
	isAlive bool
}

func GetDiffForRendering(oldGrid, newGrid Grid) []Diff {
	var diffs []Diff
	for y := 0; y < len(oldGrid); y++ {
		for x := 0; x < len(oldGrid); x++ {
			if oldGrid[y][x] != newGrid[y][x] {
				diffs = append(diffs, Diff{x, y, newGrid[y][x]})
			}
		}
	}
	return diffs
}

func RenderDiff(diff []Diff) {
	// The x axis needs to be offset by a multiple of 3, as each rendered square,
	// including the following space, is 3 characters wide. Also the terminal indices
	// are one-based, so we need to offset by 1 from the grid indices that are 0-based.
	for _, d := range diff {
		if d.isAlive {
			fmt.Printf("\033[%d;%dH🟪 ", d.y+1, (d.x*3)+1)
		} else {
			fmt.Printf("\033[%d;%dH⬜️ ", d.y+1, (d.x*3)+1)
		}
	}
}
